Tasks/NpmAuthenticate/npmauth.ts

```diff
@@ -2,112 +2,104 @@ import * as path from 'path';
-            npmrcTable['index']++;
-            fs.writeFileSync(indexFile, JSON.stringify(npmrcTable));
-            util.saveFileWithName(npmrc, npmrcTable[npmrc], saveNpmrcPath);
-        }
+        
     }
     else { //If the file doesn't exist, create it. 
         npmrcTable = new Object();
         npmrcTable['index']=0;
+    }
+
+    if (npmrcTable[npmrc]=== undefined) {
         npmrcTable[npmrc] = npmrcTable['index'];
         npmrcTable['index']++;
         fs.writeFileSync(indexFile, JSON.stringify(npmrcTable));
-        util.saveFileWithName(npmrc, '0', saveNpmrcPath);
-
+        util.saveFileWithName(npmrc, npmrcTable[npmrc], saveNpmrcPath);
     }
 
     let endpointRegistries;
-    let endpointIds = tl.getDelimitedInput(constants_1.NpmAuthenticateTaskInput.CustomEndpoint, ',');
+    let endpointIds = tl.getDelimitedInput(constants.NpmAuthenticateTaskInput.CustomEndpoint, ',');
     if (endpointIds && endpointIds.length > 0) {
-        endpointRegistries = endpointIds.map(e => npmregistry_1.NpmRegistry.FromServiceEndpoint(e, true));
+        endpointRegistries = endpointIds.map(e => npmregistry.NpmRegistry.FromServiceEndpoint(e, true));
     }
-    let LocalNpmRegistries = await util.getLocalNpmRegistries(workingDir);
+    let LocalNpmRegistries = await util.getLocalNpmRegistries(workingDirectory);
     
     
-    let npmrcFile = fs.readFileSync(npmrc, 'utf8').split('\n');
-    for (let fileLine of npmrcFile) {
-        //For each registry in the .npmrc file (every line with "registry=")
-        //If it's got a service endpoint, use that. If not, use a local registry.
-        //If it's got one of those, add the credentials into the .npmrc file
-        if (fileLine.search('registry=') > -1) {  
-            let keyAndValue = fileLine.split('=');
-            let registryURL = URL.parse(keyAndValue[1]);
-            let registry;
-            if (endpointRegistries && endpointRegistries.length > 0) {
-                for (let serviceEndpoint of endpointRegistries) {
-                    let serviceURL = URL.parse(serviceEndpoint.url);
-                    if (serviceURL.hostname == registryURL.hostname) {
-                        console.log(tl.loc("AddingEndpointCredentials", registryURL.hostname));
-                        registry = serviceEndpoint;
-                        npmrcFile = clearFileOfReferences(npmrc, npmrcFile, serviceURL);
-                        break;
-                    }
+    let npmrcFile = fs.readFileSync(npmrc, 'utf8').split(os.EOL);
+    for (let RegistryURLString of npmrcparser.GetRegistries(npmrc)) {
+        let registryURL = URL.parse(RegistryURLString);
+        let registry: npmregistry.NpmRegistry;
+        if (endpointRegistries && endpointRegistries.length > 0) {
+            for (let serviceEndpoint of endpointRegistries) {
+                let serviceURL = URL.parse(serviceEndpoint.url);
+                if (serviceURL.host == registryURL.host) {
```

Comment:
I still think we need to be comparing the nerfed registries rather than the hosts. Let's say you have a couple of service endpoints:
http://foo.com/npm/one/registry/
http://foo.com/npm/two/registry/

and if your .npmrc is set to registry=//foo.com/npm/two/registry/
it would match the first one and have incorrect credentials.